### PR TITLE
enable registration agent

### PR DESCRIPTION
**Pull request by skeeey**
Add the sub command for `multicluster-controlplane`

```
# start one multicluster controlplane server
multicluster-controlplane server

# start one multicluster controlplane server and register itself to the controlplane
multicluster-controlplane server --self-management

# only start multicluster-controlplane agent, agent includes: registration-agent and work-agent (TODO)
multicluster-controlplane agent
```

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -7,61 +7,39 @@ import (
...
-			}
-
-			stopChan := genericapiserver.SetupSignalHandler()
-			if err := options.Complete(stopChan); err != nil {
-				return err
-			}
-
-			if err := options.Validate(); err != nil {
-				return err
+		Short: "Start a multicluster controlpane",
```
Comment (clyang82) [inline cmd/server/main.go:32]: typo: `controlpane` to `controlplane`

File: hack/deploy/controlplane/external-etcd-patch.yaml [hunk 2]
```diff
@@ -9,6 +9,7 @@ spec:
       - name: controlplane
         args:
           - "/multicluster-controlplane"
+          - "server"
```
Comment (clyang82) [inline hack/deploy/controlplane/external-etcd-patch.yaml:12]: need to specify the `--self-management`

File: pkg/cmd/agent/agent.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+	"open-cluster-management.io/multicluster-controlplane/pkg/agent"
+)
+
+func NewAgent() *cobra.Command {
+	agentOptions := agent.NewAgentOptions()
+
+	cmd := &cobra.Command{
+		Use:   "agent",
+		Short: "Start a Multicluster Controlpane Agent",
```
Comment (clyang82) [inline pkg/cmd/agent/agent.go:20]: Controlpane typo

File: pkg/cmd/controller/controller.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	cliflag "k8s.io/component-base/cli/flag"
+	logsapi "k8s.io/component-base/logs/api/v1"
+	"k8s.io/component-base/version/verflag"
+)
+
+func NewController() *cobra.Command {
+	options := options.NewServerRunOptions()
+	cmd := &cobra.Command{
+		Use:   "server",
+		Short: "Start a Multicluster Controlpane Server",
```
Comment (clyang82) [inline pkg/cmd/controller/controller.go:23]: nit

File: pkg/controllers/ocmcontroller/ocmagent.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	aggregatorapiserver "k8s.io/kube-aggregator/pkg/apiserver"
+
+	clusterclient "open-cluster-management.io/api/client/cluster/clientset/versioned"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	"open-cluster-management.io/multicluster-controlplane/pkg/agent"
+)
+
+const (
+	defaultOCMAgentNamespace = "open-cluster-management-agent"
+	selfManagedClusterName   = "local-cluster"
```
Comment (clyang82) [inline pkg/controllers/ocmcontroller/ocmagent.go:23]: Do we have those constants exposed in open-cluster-management.io/api? if so, we may reference there.

**Review by clyang82 — CHANGES_REQUESTED** [review]
Overall looks good for me. Just same minor comments.

Another comments: Can we run the existing test cases for the managed cluster in e2e?

**Comment by skeeey** [comment]
> Overall looks good for me. Just same minor comments.
> 
> Another comments: Can we run the existing test cases for the managed cluster in e2e?

yes, we can, but I think there are some problems for our e2e code (e.g. we should use the hack/deploy script to deploy the controlplane in the e2e) and we also need add more test cases for e2e, so I plan to enhance this part after our function code is done

File: cmd/server/main.go (hunk 1 above)
Comment (skeeey) [reply cmd/server/main.go:32]: correct

File: hack/deploy/controlplane/external-etcd-patch.yaml (hunk 2 above)
Comment (skeeey) [reply hack/deploy/controlplane/external-etcd-patch.yaml:12]: yes, I want to make this later, I want to have a pr to enhance the deploy script

File: pkg/cmd/agent/agent.go (hunk 3 above)
Comment (skeeey) [reply pkg/cmd/agent/agent.go:20]: corrected

File: pkg/cmd/controller/controller.go (hunk 4 above)
Comment (skeeey) [reply pkg/cmd/controller/controller.go:23]: corrected

File: pkg/controllers/ocmcontroller/ocmagent.go (hunk 5 above)
Comment (skeeey) [reply pkg/controllers/ocmcontroller/ocmagent.go:23]: maybe we can put the `defaultOCMAgentNamespace` to the api repo, I will add a todo here firstly :)

File: hack/deploy/controlplane/external-etcd-patch.yaml (hunk 2 above)
Comment (clyang82) [reply hack/deploy/controlplane/external-etcd-patch.yaml:12]: You may do it once https://github.com/open-cluster-management-io/multicluster-controlplane/pull/37 is merged. There are lots of changes to simplify the deploy scripts.

**Review by clyang82 — APPROVED** [review]

**Pull request merged**